plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java

```diff
@@ -1259,7 +1259,11 @@ private InitiateMultipartUploadResult initMultipartUpload(String bucket, String
                             STATS.newInitiateMultipartUploadError();
                             if (e instanceof AmazonS3Exception s3Exception) {
                                 switch (s3Exception.getStatusCode()) {
-                                    case HTTP_FORBIDDEN, HTTP_NOT_FOUND, HTTP_BAD_REQUEST -> throw new UnrecoverableS3OperationException(bucket, key, e);
+                                    case HTTP_FORBIDDEN, HTTP_BAD_REQUEST -> throw new UnrecoverableS3OperationException(bucket, key, e);
+                                    case HTTP_NOT_FOUND -> {
+                                        throwIfFileNotFound(s3Exception);
+                                        throw new UnrecoverableS3OperationException(bucket, key, e);
```

Comment:
can HTTP_NOT_FOUND indicate something else than "not found"?

cc @alexjo2144 @electrum

Reply:
As per https://docs.aws.amazon.com/AmazonS3/latest/userguide/ErrorBestPractices.html

> HTTP status codes are less expressive than Amazon S3 error codes and contain less information about the error. For example, the NoSuchKey and NoSuchBucket Amazon S3 errors both map to the HTTP 404 Not Found status code.
Although the HTTP status codes contain less information about the error, clients that understand HTTP, but not the Amazon S3 API, will usually handle the error correctly.
Therefore, when handling errors or reporting Amazon S3 errors to end users, use the Amazon S3 error code instead of the HTTP status code as it contains the most information about the error.